File: packages/core/halo/credentials/src/credentials/verifier.ts
//
// Copyright 2022 DXOS.org
//

import { verifySignature } from '@dxos/crypto';
import { type PublicKey } from '@dxos/keys';
import { toPublicKey } from '@dxos/protocols/buf';
import { type Chain, type Credential } from '@dxos/protocols/buf/dxos/halo/credentials_pb';

import { isValidAuthorizedDeviceCredential } from './assertions';
import { getCredentialProofPayload } from './signing';

export const SIGNATURE_TYPE_ED25519 = 'ED25519Signature';

export type VerificationResult = { kind: 'pass' } | { kind: 'fail'; errors: string[] };

export const verifyCredential = async (credential: Credential): Promise<VerificationResult> => {
  const proof = credential.proof;
  if (!proof) {
    return { kind: 'fail', errors: ['Credential is missing a proof.'] };
  }
  const issuer = toPublicKey(credential.issuer);
  const signer = toPublicKey(proof.signer);
  if (!issuer || !signer) {
    return { kind: 'fail', errors: ['Credential is missing its issuer or signer.'] };
  }

  if (!issuer.equals(signer)) {
    if (!proof.chain) {
      return {
        kind: 'fail',
        errors: ['Delegated credential is missing credential chain.'],
      };
    }

    const result = await verifyChain(proof.chain, issuer, signer);
    if (result.kind === 'fail') {
      return result;
    }
  }

  const result = await verifyCredentialSignature(credential);
  if (result.kind === 'fail') {
    return result;
  }

  return { kind: 'pass' };
};

/**
 * Verifies that the signature is valid and was made by the signer.
 * Does not validate other semantics (e.g. chains).
 */
export const verifyCredentialSignature = async (credential: Credential): Promise<VerificationResult> => {
  const proof = credential.proof;
  const signer = toPublicKey(proof?.signer);
  if (!proof || !signer) {
    return { kind: 'fail', errors: ['Credential is missing a proof.'] };
  }
  if (proof.type !== SIGNATURE_TYPE_ED25519) {
    return {
      kind: 'fail',
      errors: [`Invalid signature type: ${proof.type}`],
    };
  }

  const signData = getCredentialProofPayload(credential);
  if (!(await verifySignature(signer, signData, proof.value))) {
    return { kind: 'fail', errors: ['Invalid signature'] };
  }

  return { kind: 'pass' };
};

/**
 * Verifies that the signer has the delegated authority to create credentials on behalf of the issuer.
 */
export const verifyChain = async (
  chain: Chain,
  authority: PublicKey,
  subject: PublicKey,
): Promise<VerificationResult> => {
  if (!chain.credential) {
    return { kind: 'fail', errors: ['Credential chain is empty.'] };
  }

  const result = await verifyCredential(chain.credential);
  if (result.kind === 'fail') {
    return result;
  }

  if (!isValidAuthorizedDeviceCredential(chain.credential, authority, subject)) {
    return {
      kind: 'fail',
      errors: [`Invalid credential chain: invalid assertion for key: ${subject}`],
    };
  }

  return { kind: 'pass' };
};

File: packages/core/halo/go-credentials/verifier.go
package credentials

import (
	"bytes"

	credentialspb "github.com/dxos/dxos/proto/def/dxos/halo/credentials"
	keyspb "github.com/dxos/dxos/proto/def/dxos/keys"
	"github.com/dxos/dxos/utils/keys"
)

const (
	VERIFICATION_RESULT_KIND_VALID   = "pass"
	VERIFICATION_RESULT_KIND_INVALID = "fail"
)

type VerificationResult struct {
	Kind   string
	Errors []string
}

func VerifyCredential(credential *credentialspb.Credential) (*VerificationResult, error) {
	if !bytes.Equal(credential.Issuer.Data, credential.Proof.Signer.Data) {
		if credential.Proof.Chain == nil {
			return &VerificationResult{
				Kind:   VERIFICATION_RESULT_KIND_INVALID,
				Errors: []string{"Delegated credential is missing credential chain."},
			}, nil
		}
		result, err := VerifyChain(credential.Proof.Chain, credential.Issuer, credential.Proof.Signer)
		if err != nil {
			return nil, err
		}
		if result.Kind == VERIFICATION_RESULT_KIND_INVALID {
			return result, nil
		}
	}

	return VerifyCredentialSignature(credential)
}

func VerifyCredentialSignature(credential *credentialspb.Credential) (*VerificationResult, error) {
	if credential.Proof.Type != SIGNATURE_TYPE_ED25519 {
		return &VerificationResult{
			Kind:   VERIFICATION_RESULT_KIND_INVALID,
			Errors: []string{"Invalid signature type: " + credential.Proof.Type},
		}, nil
	}

	signData, err := GetSignaturePayloadForCredential(credential)
	if err != nil {
		return nil, err
	}

	if !keys.VerifySignature(credential.Proof.Signer, signData, credential.Proof.Value) {
		return &VerificationResult{
			Kind:   VERIFICATION_RESULT_KIND_INVALID,
			Errors: []string{"Invalid signature"},
		}, nil
	}

	return &VerificationResult{
		Kind:   VERIFICATION_RESULT_KIND_VALID,
		Errors: []string{},
	}, nil
}

func VerifyChain(chain *credentialspb.Chain, authority *keyspb.PublicKey, subject *keyspb.PublicKey) (*VerificationResult, error) {
	return VerifyCredential(chain.Credential)
}
